lld/ELF/Writer.cpp

```diff
@@ -1462,6 +1462,8 @@ template <class ELFT> void Writer<ELFT>::finalizeAddressDependentContent() {
   for (;;) {
     bool changed = ctx.target->needsThunks
                        ? tc.createThunks(pass, ctx.outputSections)
+                   : ctx.arg.emachine == EM_LOONGARCH && !ctx.arg.relax
```

Comment:
Adding `ctx.arg.relax` to every place is not acceptable. You could change `ctx.arg.relax` to be always false for loongarch, though.

Reply:
> Adding `ctx.arg.relax` to every place is not acceptable. You could change `ctx.arg.relax` to be always false for loongarch, though.

ctx.arg.relax is used here because we want to determine whether or not to apply relaxation based on the parameters passed at the time of linking.

Reply:
> ctx.arg.relax is used here because we want to determine whether or not to apply relaxation based on the parameters passed at the time of linking.

I don't think so. R_LARCH_ALIGN should be handled unconditionally, otherwise the code may be unaligned.

Reply:
> > ctx.arg.relax is used here because we want to determine whether or not to apply relaxation based on the parameters passed at the time of linking.
> 
> I don't think so. R_LARCH_ALIGN should be handled unconditionally, otherwise the code may be unaligned.

Yes, it was an oversight. So for scenarios outside of align, is it necessary to determine whether or not to apply relaxation via the incoming link parameter, i.e. --relax? For example, in the relax function

Reply:
> Yes, it was an oversight. So for scenarios outside of align, is it necessary to determine whether or not to apply relaxation via the incoming link parameter, i.e. --relax? For example, in the relax function

I'm afraid it's not necessary because R_LARCH_RELAX is converted to R_NONE by function `getRelExpr` when `--no-relax`.
```
  case R_LARCH_RELAX:
    return ctx.arg.relax ? R_RELAX_HINT : R_NONE;
  case R_LARCH_ALIGN:
    return R_RELAX_HINT
```

Is it correct? @MQ-mengqing

Reply:
> > Yes, it was an oversight. So for scenarios outside of align, is it necessary to determine whether or not to apply relaxation via the incoming link parameter, i.e. --relax? For example, in the relax function
> 
> I'm afraid it's not necessary because R_LARCH_RELAX is converted to R_NONE by function `getRelExpr` when `--no-relax`.
> 
> ```
>   case R_LARCH_RELAX:
>     return ctx.arg.relax ? R_RELAX_HINT : R_NONE;
>   case R_LARCH_ALIGN:
>     return R_RELAX_HINT
> ```
> 
> Is it correct? @MQ-mengqing

Thanks for the reminder, I test both on RISCV and LoongArch, the relocation which corresponding RelExpr is R_NONE will be ignored. The changes of test files will be reverted.